src/main/java/net/mcreator/ui/minecraft/states/IPropertyData.java

```diff
@@ -0,0 +1,129 @@
+			.collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
+
+	/**
+	 * @return The name of this property.
+	 */
+	String getName();
+
+	/**
+	 * Provides the default value of type of this property. This is the "null" value of this type, which means
+	 * it may be outside value limits defined for a particular property.
+	 *
+	 * @return A default value of this property's type.
+	 */
+	@Nonnull T getDefaultValue();
+
+	/**
+	 * Converts passed value of this property to its string representation.
+	 *
+	 * @param value A value of this property's type.
+	 * @return Possible value of this property as a string.
+	 * @throws ClassCastException if the type of passed value doesn't match the type of property or its subtype.
+	 */
+	String toString(Object value);
+
+	/**
+	 * Parses string representation of passed value of this property.
+	 *
+	 * @param value Possible value of this property as JsonElement.
+	 * @return A value of this property's type.
+	 */
+	T parseObj(JsonElement value);
+
+	/**
+	 * Generates a UI component accepting values of type {@link T} and sets its value to the passed one.
+	 *
+	 * @param mcreator The future parent window of the component returned.
+	 * @param value    Possible value of this property.
+	 * @return A UI component that accepts values of type {@link T}.
+	 * @throws ClassCastException if the type of passed value doesn't match the type of property or its subtype.
+	 */
+	JComponent getComponent(MCreator mcreator, @Nullable Object value);
+
+	/**
+	 * Extracts possible value of this property from the provided UI component.
+	 *
+	 * @param component A UI component that accepts values of type {@link T}.
+	 * @return A value of this property's type.
+	 */
+	T getValue(JComponent component);
+
+	/**
+	 * We need a custom serializer/deserializer for this class because we need to store the type of property.
+	 * Technically type could be determined from properties list, but we don't have a reference to it, and it also
+	 * depends on the ME type, so this is second-best option. There is minimal overhead in storing the type.
+	 */
+	class GSONAdapter implements JsonSerializer<IPropertyData<?>>, JsonDeserializer<IPropertyData<?>> {
+
+		private static final Gson gson = new GsonBuilder().disableHtmlEscaping().setPrettyPrinting().setLenient()
+				.create();
+
+		@Override
+		public IPropertyData<?> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+				throws JsonParseException {
+			JsonObject jsonObject = json.getAsJsonObject();
+			PropertyData<?> data = gson.fromJson(jsonObject, typeMappings.get(jsonObject.get("type").getAsString()));
+			return !jsonObject.get("name").getAsString().startsWith("CUSTOM:") ? new BuiltInPropertyData<>(data) : data;
+		}
+
+		@Override
+		public JsonElement serialize(IPropertyData<?> propertyData, Type typeOfSrc, JsonSerializationContext context) {
+			JsonObject retVal = new JsonObject();
+			retVal.addProperty("name", propertyData.getName());

```

Comment:
Wanted to ask all this time: why don't we store property configuration in these objects?

Reply:
I think a better solution to this is to read datalist and load configuration from there as this is doable

Reply:
We could move logic from JItemPropertiesStatesList L67 onwards in eg. builtinproperty class and use same method here and there.

Reply:
>We could move logic from JItemPropertiesStatesList L67 onwards in eg. builtinproperty class and use same method here and there.

So should I implement this?

Reply:
Depends. If you think in the future custom properties will have min/max too, then it is better to store min max and just serialize them (in this case I can implement this as it is just one line of change + removal of transient with little GSON trick).

If this is not the case, you should implement:

> We could move logic from JItemPropertiesStatesList L67 onwards in eg. builtinproperty class and use same method here and there.

Reply:
Block state properties do have min/max values limitation, so...

Reply:
Then wait for me to "fix" this as I have good idea how to do it. But tomorrow, GTG today